### DISCUSSION TITLE
Asynchronous ADC timed readings

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#13742] [STM32 / Pyboard] [answered]
For the stm32 port it is possible to read the ADC in a timed manner. See [docs](https://docs.micropython.org/en/latest/library/pyb.ADC.html#pyb-adc). Should it be possible to use this in a non-blocking way?
@peterhinch has a very useful asynchronous implementation for reading and waiting an ADC, but this is implemented for a single read see [AADC](https://github.com/peterhinch/micropython-async/blob/master/v3/primitives/aadc.py).

**Comment by Josverl** [comment]
The docs you linked to explicitly state that the function is blocking.

> This function does not allocate any heap memory. It has blocking behaviour: it does not return to the calling program until the buffer is full.

Can you explain what you are trying to achieve?

**Reply by saraverbeecke** [reply]
Hi thanks for looking into this. Indeed it says blocking so a no-go I suppose.
I would like to read out some values at a certain sample rate (1kHz) but need to be able to await this so I can do other stuff while the ADC values are being buffered...

**Reply by peterhinch** [reply, marked as answer]
1KHz is too fast for an `asyncio` solution. One approach would be to do the ADC read in a hard ISR. [The micropython-filters](https://github.com/peterhinch/micropython-filters) repo illustrates realtime data capture on STM32. The [lpf.py demo](https://github.com/peterhinch/micropython-filters/blob/master/lpf.py) uses a timer ISR to capture data.

**Reply by saraverbeecke** [reply]
OK thanks for the info. The LPF demo looks great, I'll try that approach on my stm32f411.

**Reply by peterhinch** [reply]
Just to clarify, your application can use `asyncio` so long as your data capture is by ISR.
